field/TinyLang/Field/Generator.hs

```diff
@@ -269,6 +201,51 @@ arbitraryUnOpRing = elements [Neg]
 arbitraryBinOpRing :: MonadGen m => m (BinOp f (AField f) (AField f) (AField f))
 arbitraryBinOpRing = elements [Add, Sub, Mul]
 
+-- | Generate an expression of a particular type from a collection of variables
+-- with the number of nodes (approximately) bounded by 'size'.
+boundedArbitraryExpr
+    :: forall f a m. (Field f, Arbitrary f, KnownUni f a, MonadGen m, MonadSupply m)
+    => Vars f -> Int -> m (Expr f a)
+boundedArbitraryExpr _    size | size <= 1 = EVal <$> arbitraryM
+boundedArbitraryExpr vars size             = frequency $ everything ++ comparisons where
+    -- The most general generator.
+    everything =
+        [ (1, EVal <$> arbitraryM)
+        , (2, EVar <$> chooseUniVar vars)
+        , (2, do
+                let size' = size `Prelude.div` 3
+                EIf
+                    <$> boundedArbitraryExpr vars size'
+                    <*> boundedArbitraryExpr vars size'
+                    <*> boundedArbitraryExpr vars size')
+        , (4, withOneofUnis $ \(_ :: Uni f a') -> do
+                uniVar <- genFreshUniVar @f @a'
+                let vars' = Some uniVar : vars
+                    size' = size `Prelude.div` 2
+                ELet uniVar
+                    <$> boundedArbitraryExpr vars  size'
+                    <*> boundedArbitraryExpr vars' size')
+        , (4, withOneofUnOps  $ \unOp  -> do
+                let size' = size - 1
+                EAppUnOp unOp <$> boundedArbitraryExpr vars size')
+        , (4, withOneofBinOps $ \binOp -> do
+                let size' = size `Prelude.div` 2
+                EAppBinOp binOp
+                    <$> boundedArbitraryExpr vars size'
+                    <*> boundedArbitraryExpr vars size')
+        ]
```

Comment:
Do we need a case for EConstr yet?

Reply:
Yes, but let's do that separately. If done naively, we'll get unsatisfied constraints most of the time, while we want the opposite.